tests/unit/test_charm.py

```diff
@@ -86,6 +86,20 @@ def test_given_private_key_is_stored_and_certificate_relation_is_created_when_on
 
         patch_request_certificate.assert_called_with(certificate_signing_request=csr.encode())
 
+    @patch(
+        "charms.tls_certificates_interface.v2.tls_certificates.TLSCertificatesRequiresV2.request_certificate_creation"  # noqa: E501, W505
+    )
+    @patch("charm.generate_csr")
+    def test_given_csr_is_not_generated_and_certificate_relation_is_created_then_request_certificate_is_not_called(  # noqa: E501
+        self,
+        patch_generate_csr,
+        patch_request_certificate,
+    ):
+        self.harness.set_leader(is_leader=True)
+        patch_generate_csr.side_effect = RuntimeError("Private key not stored.")
+        self.harness.add_relation(relation_name="certificates", remote_app="certificates-provider")
+        patch_request_certificate.assert_not_called()
```

Comment:
I don't really understand how this relates to the change made.

The change is in the behaviour of `_request_certificate()`, but the test is asserting that it isn't called.

Reply:
Here patch_request_certificate 
mocking "charms.tls_certificates_interface.v2.tls_certificates.TLSCertificatesRequiresV2.request_certificate_creation" which is  called in _request_certificate. I think changing mock name form patch_request_certificate to patch_request_certificate_creation makes prevents the confusion here.

Reply:
Done

Reply:
Oh, yes, that had me a little confused. Thanks for the clarification there!

Admittedly, I'm still a little confused about this test, though.

The test says to me "if `_generate_csr` throws an exception, then when `_request_certificates()` is called ,we won't call `_request_certificate_relation()`"

There's two places where `_request_certificates()` is called.

1. in `_on_config_changed()` where `_generate_csr()` is called beforehand.Iit wouldn't be called, since the exception would halt the execution before it gets there.

2. in `_on_certificates_relation_joined()`, where there is already a `if not self._csr_is_stored` check before calling it, and it so it returns before calling it

So it's just not clear to me what this is testing. If you remove your change, does this test fail?

Reply:
Hi,
I updated the unit test which validates the change explicitly. Please check the final version. Thank you.